### DISCUSSION TITLE
Which component handles the insert operation?

### DISCUSSION THREAD

**Discussion by ArenaSu** [milvus-io/milvus#33426] [Q&A and General discussion]
Which component handles the insert operation (internal/distributed/proxy/httpserver/handler.go handleInsert) after proxy send msg to queue?What rules for naming message queue topics?

**Comment by yhmo** [comment]
When you insert some data by insert():
1. The data is received by proxy node, proxy node splits the data into pieces by shards_num(which is defined in collection schema, default is 1), and sends the pieces(as messages) into Pulsar/Kafka. The insert() returns success to the client once the data is persisted into Pulsar/Kafka.
2. For each growing segment, two "data channels" are created, one is between querynode and Pulsar/Kafka, the other is between datanode and Pulsar/Kafka.
3. Querynode and datanode subscribe messages asynchronously from Pulsar/Kafka, the data pieces are consumed by querynode and datanode.
4. For each growing segment, datanode keeps a buffer to accept new data. Once the buffer size reaches 16MB or the buffer lifetime hits the limit(10 minutes), the buffer is persisted to be a chunk into S3/MinIO.
5. If the collection is loaded, all segments are kept in query nodes. for each growing segment, querynode keeps a buffer to accept new data, so that search requests can search the data. New data is consumed by querynode from Pulsar/Kafka, only when the new data is already consumed by querynode, the data is visible&searchable by search()/query(). 
6. With more and more chunks persisted, the size of the growing segment reaches a threshold(100+MB), the growing segment is converted to be sealed segment by data coordinator. No more data is allowed to be inserted into this segment. Query node and data node will create a new buffer for the next growing segment.
7. Once a new sealed segment is generated, index node will receive a task to build index for this seal segment.
8. Once an index is finished, the query node will load this index into memory to replace the original segment data.

**Reply by ArenaSu** [reply]
Can you tell me the function name in querynode and datanode to handle insert api as i am reading the source code ?

**Reply by yhmo** [reply]
For datanode:
internal/datanode/data_sync_service.go in the function getServiceWithChannel()
`dmStreamNode, err := newDmInputNode(initCtx, node.dispClient, info.GetVchan().GetSeekPosition(), config)`
The dmStreamNode is to consume data from pulsar.
newDmInputNode() is in this file:
internal/datanode/flow_graph_dmstream_input_node.go


For querynode:
internal/querynodev2/pipeline/pipeline.go the function NewPipeLine() construct a pipeline to consume data from pulsar.
The indexNode is for consuming insert data, which is implemented in internal/querynodev2/pipeline/insert_node.go

**Reply by ArenaSu** [reply]
For datanode:
In internal/datanode/event_manager.go StartWatchChannels->handleChannelEvt->handleWatchInfo->handlePutEvent. Is this right？

**Reply by yhmo** [reply]
I am not familiar with the source code details of this area. You need to read it by yourself.

**Reply by ArenaSu** [reply]
> For datanode: In internal/datanode/event_manager.go StartWatchChannels->handleChannelEvt->handleWatchInfo->handlePutEvent. Is this right？

@xiaofan-luan Could you please check if it is correct ？
